### ISSUE TITLE
Avoid 400 on initial check-in -attempt

### ISSUE THREAD

**Issue by studgeek** [jdholtz/auto-southwest-check-in#325]
### Description

Per @jdholtz, a 400 at check-in usually means the attempt was too early (as it went through in later attempts). This can happen on the first attempt as the time isn’t exactly synchronized between the Southwest servers and NTP servers.

One experiment could be checking in a second or two after the minute to see if we can avoid this first attempt failure. This may then speed up the check-in as it takes ~4 seconds for the first attempt to report a failure before the script tries again.

**Comment by studgeek** [comment]
Here is example of 400 on first attempt by the script...

```
$ pipenv run python3 southwest.py XXXXXX My Name -v
2024-12-21 11:28:31 DEBUG MainProcess[log:24]: Initialized the application
2024-12-21 11:28:31 DEBUG MainProcess[main:113]: Auto-Southwest Check-In v8.1
2024-12-21 11:28:31 DEBUG MainProcess[main:71]: Called with 3 arguments
2024-12-21 11:28:31 DEBUG MainProcess[config:143]: Initializing configuration file
2024-12-21 11:28:31 DEBUG MainProcess[config:172]: Reading the configuration file
2024-12-21 11:28:31 DEBUG MainProcess[config:176]: No configuration file found. Using defaults
2024-12-21 11:28:31 DEBUG MainProcess[config:185]: Reading configuration from environment variables
2024-12-21 11:28:31 DEBUG MainProcess[main:84]: Adding reservation through CLI arguments
...
2024-12-21 17:25:05 DEBUG Process-1:1[checkin_handler:119]: Sleeping until check-in: 1794 seconds...
Checking in to flight from 'Oakland' to 'Phoenix' for My Name

2024-12-21 17:55:00 DEBUG Process-1:1[checkin_handler:169]: Attempting to check in
2024-12-21 17:55:00 DEBUG Process-1:1[checkin_handler:204]: Making first POST request to check in
2024-12-21 17:55:04 DEBUG Process-1:1[utils:90]: Request error on attempt 1: Bad Request (400). Sleeping for 0.50 seconds until next attempt
2024-12-21 17:55:06 DEBUG Process-1:1[utils:70]: Successfully made request after 2 attempts
2024-12-21 17:55:06 DEBUG Process-1:1[checkin_handler:211]: Making second POST request to check in
2024-12-21 17:55:08 DEBUG Process-1:1[utils:70]: Successfully made request after 1 attempts
2024-12-21 17:55:08 DEBUG Process-1:1[checkin_handler:179]: Successfully checked in after 1 attempts
2024-12-21 17:55:08 DEBUG Process-1:1[notification_handler:119]: Sending successful check-in notification...
```

Version info:
```
M$ git l2 -1
* 2fde441 - Sun, 24 Nov 2024 20:00:00 -0600 (4 weeks ago) (HEAD, origin/develop)
|           Update changelog to include JSON schema addition - jdholtz
```

**Comment by studgeek** [comment]
I tried adding just a second as follows...
```
--- a/lib/checkin_handler.py
+++ b/lib/checkin_handler.py
@@ -78,7 +78,7 @@ class CheckInHandler:

     def _set_check_in(self) -> None:
         # Check-in is 24 hours before the flight departs
-        checkin_time = self.flight.departure_time - timedelta(days=1)
+        checkin_time = self.flight.departure_time - timedelta(days=1) + timedelta(seconds=1)
```

I still had a 400 though. And resulting seats were not great...
```
2024-12-29 16:20:05 DEBUG Process-1:1[checkin_handler:119]: Sleeping until check-in: 1795 seconds...
Checking in to flight from 'Phoenix' to 'Oakland' for David Rees

2024-12-29 16:50:01 DEBUG Process-1:1[checkin_handler:169]: Attempting to check in
2024-12-29 16:50:01 DEBUG Process-1:1[checkin_handler:208]: Making first POST request to check in
2024-12-29 16:50:05 DEBUG Process-1:1[utils:91]: Request error on attempt 1: Bad Request (400). Sleeping for 0.20 seconds until next attempt
2024-12-29 16:50:11 DEBUG Process-1:1[utils:70]: Successfully made request after 2 attempts
2024-12-29 16:50:11 DEBUG Process-1:1[checkin_handler:215]: Making second POST request to check in
2024-12-29 16:50:14 DEBUG Process-1:1[utils:70]: Successfully made request after 1 attempts
2024-12-29 16:50:14 DEBUG Process-1:1[checkin_handler:183]: Successfully checked in after 1 attempts
2024-12-29 16:50:14 DEBUG Process-1:1[notification_handler:119]: Sending successful check-in notification...
```

**Comment by studgeek** [comment]
Could try adding another second.

Or, I wonder if we could have the script run multiple tabs (like I sometimes do manually in the browser). For example, it could start checkin in tab 1 at exact checkin time, tab 2 at +1 seconds, and tab 3 at +2 seconds.

**Comment by jdholtz** [comment]
> For example, it could start checkin in tab 1 at exact checkin time, tab 2 at +1 seconds, and tab 3 at +2 seconds.

That's a good idea. I can come up with a proof of concept for this so we can evaluate how well it does in practice. Do you have a check-in coming up soon?

**Comment by dmytrokoren** [comment]
> > For example, it could start checkin in tab 1 at exact checkin time, tab 2 at +1 seconds, and tab 3 at +2 seconds.
> 
> 
> 
> That's a good idea. I can come up with a proof of concept for this so we can evaluate how well it does in practice. Do you have a check-in coming up soon?

I have a check-in coming up soon. Will you be available to do POC?

**Comment by jdholtz** [comment]
> I have a check-in coming up soon. Will you be available to do POC?

I'll have some time this weekend. Will that be too late for your check-in?

**Comment by dmytrokoren** [comment]
> > I have a check-in coming up soon. Will you be available to do POC?
> 
> I'll have some time this weekend. Will that be too late for your check-in?

March 9th is the check in

**Comment by jdholtz** [comment]
@dmytrokoren I've added a potentially working version at the [multiple-checkin-threads](https://github.com/jdholtz/auto-southwest-check-in/tree/multiple-checkin-threads) branch (tested manually, but not yet with an actual flight check-in). This will start a separate thread every second from 0 to 4 seconds after the check-in time.

A potential concern that came up during this development is that Southwest _could_ more easily detect/block users because multiple requests for check-in are being received (potentially up to 5x). However, I don't think this is much of an issue as it is common for people to use multiple devices/tabs to check-in at once (an example being stdugeek doing this manually with multiple browser tabs) so this wouldn't be out of the ordinary.

**Comment by jdholtz** [comment]
@dmytrokoren has tested the multiple-checkin-threads branch here: https://github.com/jdholtz/auto-southwest-check-in/commit/0afb60f68ce00de1025a433793aecc899683fad9#commitcomment-153482366

Here's my response copied over:
> Awesome, thanks for testing. I unfortunately didn't add a way to differentiate logs between different threads so we couldn't see if the threads starting later actually checked in earlier (having avoided the 400). I just added a commit that will now put a thread ID in the log message so we can differentiate. We should do a couple more check-ins to see if this strategy actually works better.

TL;DR: we should run a few more tests to see if staggering the check-in times by a few seconds does improve the check-in speed.

**Comment by jdholtz** [comment]
Here are results from a flight earlier today that I ran with the `multiple-checkin-threads` branch. The second thread, being the one delayed by 1 second, checked in first. I'd like to get a few more runs before solidifying a delay of 1 though.

<details>
<summary>Logs</summary>

```
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:156]: Starting thread with sleep time 0
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:156]: Starting thread with sleep time 1
2025-07-22 13:49:59 DEBUG Process-1:1:CheckInThread-1[checkin_handler:226]: Attempting to check in
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:156]: Starting thread with sleep time 2
2025-07-22 13:49:59 DEBUG Process-1:1:CheckInThread-1[checkin_handler:265]: Making first POST request to check in
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:156]: Starting thread with sleep time 3
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:156]: Starting thread with sleep time 4
2025-07-22 13:49:59 DEBUG Process-1:1[checkin_handler:164]: All threads started. Waiting for them to finish
2025-07-22 13:50:00 DEBUG Process-1:1:CheckInThread-2[checkin_handler:226]: Attempting to check in
2025-07-22 13:50:00 DEBUG Process-1:1:CheckInThread-2[checkin_handler:265]: Making first POST request to check in
2025-07-22 13:50:01 DEBUG Process-1:1:CheckInThread-3[checkin_handler:226]: Attempting to check in
2025-07-22 13:50:01 DEBUG Process-1:1:CheckInThread-3[checkin_handler:265]: Making first POST request to check in
2025-07-22 13:50:02 DEBUG Process-1:1:CheckInThread-4[checkin_handler:226]: Attempting to check in
2025-07-22 13:50:02 DEBUG Process-1:1:CheckInThread-4[checkin_handler:265]: Making first POST request to check in
2025-07-22 13:50:03 DEBUG Process-1:1:CheckInThread-5[checkin_handler:226]: Attempting to check in
2025-07-22 13:50:03 DEBUG Process-1:1:CheckInThread-5[checkin_handler:265]: Making first POST request to check in
2025-07-22 13:50:04 DEBUG Process-1:1:CheckInThread-1[utils:101]: Request error on attempt 1: Bad Request (400). Sleeping for 0.50 seconds until next attempt
2025-07-22 13:50:06 DEBUG Process-1:1:CheckInThread-2[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:06 DEBUG Process-1:1:CheckInThread-2[checkin_handler:272]: Making second POST request to check in
2025-07-22 13:50:07 DEBUG Process-1:1:CheckInThread-3[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:07 DEBUG Process-1:1:CheckInThread-3[checkin_handler:272]: Making second POST request to check in
2025-07-22 13:50:07 DEBUG Process-1:1:CheckInThread-1[utils:81]: Successfully made request after 2 attempts
2025-07-22 13:50:07 DEBUG Process-1:1:CheckInThread-1[checkin_handler:272]: Making second POST request to check in
2025-07-22 13:50:08 DEBUG Process-1:1:CheckInThread-4[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:08 DEBUG Process-1:1:CheckInThread-4[checkin_handler:272]: Making second POST request to check in
2025-07-22 13:50:09 DEBUG Process-1:1:CheckInThread-5[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:09 DEBUG Process-1:1:CheckInThread-5[checkin_handler:272]: Making second POST request to check in
2025-07-22 13:50:09 DEBUG Process-1:1:CheckInThread-2[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:09 DEBUG Process-1:1:CheckInThread-2[checkin_handler:240]: Successfully checked in after 1 attempts
2025-07-22 13:50:10 DEBUG Process-1:1:CheckInThread-3[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:10 DEBUG Process-1:1:CheckInThread-3[checkin_handler:240]: Successfully checked in after 1 attempts
2025-07-22 13:50:10 DEBUG Process-1:1:CheckInThread-1[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:10 DEBUG Process-1:1:CheckInThread-1[checkin_handler:240]: Successfully checked in after 1 attempts
2025-07-22 13:50:11 DEBUG Process-1:1:CheckInThread-5[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:11 DEBUG Process-1:1:CheckInThread-5[checkin_handler:240]: Successfully checked in after 1 attempts
2025-07-22 13:50:11 DEBUG Process-1:1:CheckInThread-4[utils:81]: Successfully made request after 1 attempts
2025-07-22 13:50:11 DEBUG Process-1:1:CheckInThread-4[checkin_handler:240]: Successfully checked in after 1 attempts
2025-07-22 13:50:11 DEBUG Process-1:1[checkin_handler:167]: All threads completed. Checking their results
```
</details>

**Comment by studgeek** [comment]
One thought on the changes in multiple-checkin-threads branch is that instead of joining the threads then iterating through the Queue. It could go right to doing a blocking get (`get(true)`) on the Queue so that it gets each result as soon as they are added. Once there is a checkin it can kill the other threads to avoid extra attempts (to reduce chance of Southwest flagging).

**Comment by jdholtz** [comment]
Yeah, there are definitely optimizations that can be done for that. However, my goal with that branch was just to determine what delay should be added to the check-in to avoid the 400. I don't think having multiple check-in threads would be really any improvement over just delaying 1 or 2 seconds.

That said, from my previous logs it looks like avoiding this 400 would lead to at most a 0.5-1 second faster check-in, so the benefits won't be too great.

**Comment by jdholtz** [comment]
Another run for more data/analysis. On this one, the first, second, and third threads all checked in within the same second (the third being the first). All threads except the second one ran into a 429, which I assume is very similar to a 400. Overall, I think this run also shows that a 1 second delay could be ideal to avoid the initial 400.

<details>
<summary>Logs</summary>

```
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:161]: Starting thread with sleep time 0
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:161]: Starting thread with sleep time 1
2025-09-20 22:25:00 DEBUG Process-1:1:CheckInThread-1[checkin_handler:231]: Attempting to check in
2025-09-20 22:25:00 DEBUG Process-1:1:CheckInThread-1[checkin_handler:270]: Making first POST request to check in
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:161]: Starting thread with sleep time 2
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:161]: Starting thread with sleep time 3
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:161]: Starting thread with sleep time 4
2025-09-20 22:25:00 DEBUG Process-1:1[checkin_handler:169]: All threads started. Waiting for them to finish
2025-09-20 22:25:01 DEBUG Process-1:1:CheckInThread-2[checkin_handler:231]: Attempting to check in
2025-09-20 22:25:01 DEBUG Process-1:1:CheckInThread-2[checkin_handler:270]: Making first POST request to check in
2025-09-20 22:25:02 DEBUG Process-1:1:CheckInThread-3[checkin_handler:231]: Attempting to check in
2025-09-20 22:25:02 DEBUG Process-1:1:CheckInThread-3[checkin_handler:270]: Making first POST request to check in
2025-09-20 22:25:03 DEBUG Process-1:1:CheckInThread-4[checkin_handler:231]: Attempting to check in
2025-09-20 22:25:03 DEBUG Process-1:1:CheckInThread-4[checkin_handler:270]: Making first POST request to check in
2025-09-20 22:25:03 DEBUG Process-1:1:CheckInThread-4[utils:73]: Request error on attempt 1: Too Many Requests (429). Sleeping for 0.50 seconds until next attempt
2025-09-20 22:25:03 DEBUG Process-1:1:CheckInThread-3[utils:54]: Successfully made request after 1 attempts
2025-09-20 22:25:03 DEBUG Process-1:1:CheckInThread-3[checkin_handler:277]: Making second POST request to check in
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-5[checkin_handler:231]: Attempting to check in
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-5[checkin_handler:270]: Making first POST request to check in
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-3[utils:73]: Request error on attempt 1: Too Many Requests (429). Sleeping for 0.50 seconds until next attempt
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-5[utils:73]: Request error on attempt 1: Too Many Requests (429). Sleeping for 0.50 seconds until next attempt
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-1[utils:54]: Successfully made request after 1 attempts
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-1[checkin_handler:277]: Making second POST request to check in
2025-09-20 22:25:04 DEBUG Process-1:1:CheckInThread-1[utils:73]: Request error on attempt 1: Too Many Requests (429). Sleeping for 0.50 seconds until next attempt
2025-09-20 22:25:05 DEBUG Process-1:1:CheckInThread-4[utils:54]: Successfully made request after 2 attempts
2025-09-20 22:25:05 DEBUG Process-1:1:CheckInThread-4[checkin_handler:277]: Making second POST request to check in
2025-09-20 22:25:05 DEBUG Process-1:1:CheckInThread-2[utils:54]: Successfully made request after 1 attempts
2025-09-20 22:25:05 DEBUG Process-1:1:CheckInThread-2[checkin_handler:277]: Making second POST request to check in
2025-09-20 22:25:05 DEBUG Process-1:1:CheckInThread-4[utils:73]: Request error on attempt 1: Too Many Requests (429). Sleeping for 0.50 seconds until next attempt
2025-09-20 22:25:06 DEBUG Process-1:1:CheckInThread-5[utils:54]: Successfully made request after 2 attempts
2025-09-20 22:25:06 DEBUG Process-1:1:CheckInThread-5[checkin_handler:277]: Making second POST request to check in
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-3[utils:54]: Successfully made request after 2 attempts
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-3[checkin_handler:245]: Successfully checked in after 1 attempts
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-1[utils:54]: Successfully made request after 2 attempts
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-1[checkin_handler:245]: Successfully checked in after 1 attempts
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-2[utils:54]: Successfully made request after 1 attempts
2025-09-20 22:25:07 DEBUG Process-1:1:CheckInThread-2[checkin_handler:245]: Successfully checked in after 1 attempts
2025-09-20 22:25:08 DEBUG Process-1:1:CheckInThread-4[utils:54]: Successfully made request after 2 attempts
2025-09-20 22:25:08 DEBUG Process-1:1:CheckInThread-4[checkin_handler:245]: Successfully checked in after 1 attempts
2025-09-20 22:25:08 DEBUG Process-1:1:CheckInThread-5[utils:54]: Successfully made request after 1 attempts
2025-09-20 22:25:08 DEBUG Process-1:1:CheckInThread-5[checkin_handler:245]: Successfully checked in after 1 attempts
2025-09-20 22:25:08 DEBUG Process-1:1[checkin_handler:172]: All threads completed. Checking their results
2025-09-20 22:25:08 DEBUG Process-1:1[checkin_handler:184]: A successful check-in was received
```
</details>

**Comment by jdholtz** [comment]
Recently I've been getting 429s when using the checkin-threads branch to check-in. Have other people been seeing 429s when just using one thread (v9.0 or develop)? I wonder if this is actually Southwest rate limiting the check-in requests (as that is what the 429 does mean in practice).
